crates/starknet-server/tests/common/mod.rs

```diff
@@ -1,53 +1,78 @@
-const MIN_PORT: u16 = 1025;
-const MAX_PORT: u16 = 65_535;
-const SEED: usize = 42;
-const ACCOUNTS: usize = 3;
+#[cfg(test)]
+pub mod util {
+    use std::net::TcpListener;
+    use std::process::{Child, Command, Stdio};
+    use std::{thread, time};
+
+    use hyper::{Client, StatusCode, Uri};
+    use lazy_static::lazy_static;
+    use starknet_rs_providers::jsonrpc::HttpTransport;
+    use starknet_rs_providers::JsonRpcClient;
+    use thiserror::Error;
+    use tokio::sync::Mutex;
+    use url::Url;
 
-const MAX_RETRIES: usize = 10;
+    use super::constants::{ACCOUNTS, HOST, MAX_PORT, MAX_RETRIES, MIN_PORT, SEED};
+
+    #[derive(Error, Debug)]
+    pub enum TestError {
+        #[error("No free ports")]
+        NoFreePorts,
+    }
 
-fn get_free_port_listener() -> Result<u16, TestError> {
-    for port in MIN_PORT..=MAX_PORT {
-        if let Ok(listener) = TcpListener::bind(("127.0.0.1", port)) {
-            return Ok(listener.local_addr().expect("No local addr").port());
+    fn get_free_port_listener() -> Result<u16, TestError> {
+        for port in MIN_PORT..=MAX_PORT {
+            if let Ok(listener) = TcpListener::bind(("127.0.0.1", port)) {
+                return Ok(listener.local_addr().expect("No local addr").port());
+            }
+            // otherwise port is occupied
         }
-        // otherwise port is occupied
+        Err(TestError::NoFreePorts)
     }
-    Err(TestError::NoFreePorts)
-}
 
-pub(crate) struct BackgroundDevnet {
-    client: Client<HttpConnector>,
-    process: Option<Child>,
-}
+    lazy_static! {
+        /// This is to prevent TOCTOU errors; i.e. one background devnet might find one
+        /// port to be free, and while it's trying to start listening to it, another instance
+        /// finds that it's free and tries occupying it
+        static ref BACKGROUND_DEVNET_MUTEX: Mutex<()> = Mutex::new(());
+    }
 
-impl BackgroundDevnet {
-    pub(crate) fn new() -> Self {
-        BackgroundDevnet { client: Client::new(), process: None }
+    pub struct BackgroundDevnet {
+        pub json_rpc_client: JsonRpcClient<HttpTransport>,
+        process: Child,
     }
 
-    /// Ensures the background instance spawns at a free port, checks at most `MAX_RETRIES` times
-    pub(crate) async fn spawn(&mut self) -> Result<(), TestError> {
-        let free_port = get_free_port_listener().expect("No free ports");
+    impl BackgroundDevnet {
+        /// Ensures the background instance spawns at a free port, checks at most `MAX_RETRIES`
+        /// times
+        pub(crate) async fn spawn() -> Self {
+            let _mutex_guard = BACKGROUND_DEVNET_MUTEX.lock().await;
 
-        self.process = Some(
-            Command::new("cargo")
+            let free_port = get_free_port_listener().expect("No free ports");
+
+            let devnet_url = format!("http://{HOST}:{free_port}");
+            let devnet_rpc_url =
+                Url::parse(format!("{}/rpc", devnet_url.as_str()).as_str()).unwrap();
```

Comment:
We can add variant in Error enum implements url::error

```
#[error("Url error")]
UrlError(#[from] url::ParseError)
```

smth like this

then it shall be possible to do just

```
let devnet_rpc_url =
                 Url::parse(format!("{}/rpc", devnet_url.as_str()).as_str())?
```

the same is for other `unwrap`/`expect`

Reply:
Applied in https://github.com/0xSpaceShard/starknet-devnet-rs/pull/64/commits/748e6d07f014873ed72c0c8cb02e1eeb9bf8b17f